### PR TITLE
Add pull provider documentation

### PR DESCRIPTION
**Pull request by jeremylenz**
* Add `3.2. Configuring Remote Execution for Pull Client` to the Installing {Smartproxy} Server guide
* Update `3.8. Migrating From Katello Agent to Remote Execution` in the Managing Hosts guide
* Add `10.8. Setting Up the Remote Execution Pull Client` to the Managing Hosts guide

* [x] I am familiar with the [contributing](https://github.com/theforeman/foreman-documentation/blob/master/CONTRIBUTING.md) guidelines.

Please cherry-pick my commits into:

* [x] Foreman 3.4/Katello 4.6
* [ ] Foreman 3.3/Katello 4.5
* [ ] Foreman 3.2/Katello 4.4
* [ ] Foreman 3.1/Katello 4.3
* For Foreman 3.0 or older, please create a separate PR.
* We do not accept PRs for Foreman 2.3 or older.

### CODE REVIEW

File: guides/common/modules/proc_setting-up-pull-client.adoc [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+For {SmartProxies} in `pull-mqtt` mode, hosts can subscribe to remote jobs using the remote execution pull client.
+No SSH connection is required.
+
+Use this procedure to configure a host to use the pull client:
+
+.Prerequisites
+* The host is registered to {ProjectName}.
+* The host's {SmartProxy} is in `pull-mqtt` mode.
+* The {Team} {project-client-name} repository is enabled and synchronized on {ProjectServer}, and enabled on the host.
+* The host is able to communicate with its {SmartProxy} over MQTT using port 1883.
```
Comment (ehelms) [inline guides/common/modules/proc_setting-up-pull-client.adoc:13]: While it should be obvious, we also should make note that the host must be able to communicate with the {SmartProxy} over HTTP.

**Comment by jeremylenz** [comment]
@adamruzicka would also appreciate your feedback here for technical accuracy :)

File: guides/common/assembly_configuring-and-setting-up-remote-jobs.adoc [hunk 2]
```diff
@@ -16,6 +16,8 @@ include::modules/proc_configuring-the-global-smartproxy-remote-execution-setting
 
 include::modules/proc_configuring-an-alternative-directory-to-execute-remote-jobs-on-clients.adoc[leveloffset=+1]
 
+include::modules/proc_setting-up-pull-client.adoc[leveloffset=+1]
```
[the remark this answers is not in the document]
Comment (jeremylenz) [reply guides/common/assembly_configuring-and-setting-up-remote-jobs.adoc:19]: hmm, The PR preview does show them in both! maybe it's just something with my local setup.

File: guides/common/modules/proc_setting-up-pull-client.adoc [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+* The host's {SmartProxy} must be in `pull-mqtt` mode.
+* The {Team} {project-client-name} repository must be enabled and synchronized on {ProjectServer}, and enabled on the host.
+
+.Procedure
+. On the host, install the `katello-pull-transport-migrate` RPM package:
++
+----
+# yum install katello-pull-transport-migrate
+----
+The host's `subscription-manager` configuration is used to configure the `yggdrasil` client on the host, and the worker is started.
```
[the remark this answers is not in the document]
Comment (jeremylenz) [reply guides/common/modules/proc_setting-up-pull-client.adoc]: @adamruzicka @ezr-ondrej

[the remark this answers is not in the document]
Comment (jeremylenz) [reply guides/common/modules/proc_setting-up-pull-client.adoc]: @ehelms ?

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+The mode is configured per {SmartProxy}.
+Some {SmartProxies} may be configured to use `pull-mqtt` mode while others use `ssh`.
+If this is the case, it is possible that one remote job on a given host will use the pull client, and the next job on the same host will use SSH.
+If you wish to avoid this scenario, configure all {SmartProxies} to use the same mode.
+
+After the pull-based transport is set up on each {SmartProxy}, it must also be configured on each host.
+See Configuring_and_Setting_up_Remote_Jobs_admin[Configuring and Setting Up Remote Jobs] in the _Managing Hosts_ guide.
+
+NOTE: The `pull-mqtt` mode works only with the Script provider.
+Ansible and other providers will continue to use SSH.
```
Comment (adamruzicka) [inline guides/common/modules/proc_configuring-pull-transport.adoc:27]: *puts on pedantic hat*

This is somewhat misleading. Other providers I'm aware of are ansible and salt. Ansible uses ssh by default, but can be configured to use other transports per host by setting `ansible_connection` variable. Salt uses a transport based on zeromq by default.

File: guides/common/modules/proc_distributing-keys-for-remote-execution.adoc [hunk 5]
```diff
@@ -1,7 +1,9 @@
 [id="Distributing_SSH_Keys_for_Remote_Execution_{context}"]
 = Distributing SSH Keys for Remote Execution
 
-To use SSH keys for authenticating remote execution connections, you must distribute the public SSH key from {SmartProxy} to its attached hosts that you want to manage.
+For {SmartProxies} in `ssh` mode, remote execution connections may be authenticated using SSH.
```
Comment (adamruzicka) [inline guides/common/modules/proc_distributing-keys-for-remote-execution.adoc:4]: "may be"?

**Review by adamruzicka — COMMENTED** [review]
In general looks good, left some nitpicks

File: guides/common/modules/proc_distributing-keys-for-remote-execution.adoc (hunk 5 above)
Comment (jeremylenz) [reply guides/common/modules/proc_distributing-keys-for-remote-execution.adoc:4]: That was because of the sentence below
```
If the hosts are running on Amazon Web Services, enable password authentication.
```

File: guides/common/modules/proc_configuring-pull-transport.adoc (hunk 4 above)
Comment (jeremylenz) [reply guides/common/modules/proc_configuring-pull-transport.adoc:27]:
```
Ansible and other providers will continue to use their default transport settings.
```
Is that better?

Comment (adamruzicka) [reply guides/common/modules/proc_configuring-pull-transport.adoc:27]: Sounds good

File: guides/common/modules/proc_distributing-keys-for-remote-execution.adoc (hunk 5 above)
Comment (adamruzicka) [reply guides/common/modules/proc_distributing-keys-for-remote-execution.adoc:4]: Unless amazon is doing something odd in there, the authentication will always be handled by ssh, no matter if you use a public key or password based auth, no?

Comment (jeremylenz) [reply guides/common/modules/proc_distributing-keys-for-remote-execution.adoc:4]: ahh gotcha, I was just a bit confused and thought "password authentication" implied non-SSH somehow. Will update

**Comment by jeremylenz** [comment]
@adamruzicka Thanks, updated 👍

File: guides/common/modules/proc_setting-up-pull-client.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (adamruzicka) [reply guides/common/modules/proc_setting-up-pull-client.adoc]: Afaik it will still be `yggdrasil`

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+ # {installer-scenario-smartproxy} \
+  --foreman-proxy-plugin-remote-execution-script-mode pull-mqtt
+
+----
+
+The mode is configured per {SmartProxy}.
+Some {SmartProxies} may be configured to use `pull-mqtt` mode while others use `ssh`.
+If this is the case, it is possible that one remote job on a given host will use the pull client, and the next job on the same host will use SSH.
+If you wish to avoid this scenario, configure all {SmartProxies} to use the same mode.
```
[the remark this answers is not in the document]
Comment (pmoravec) [reply guides/common/modules/proc_configuring-pull-transport.adoc:21]: A big +1 for such recommendation (with a text like "Warning: In pull-mqtt mode, clients get subscribed to the SmartProxy they are registered through for job notifications. It is recommended to flip (that setting) to true to ensure the notification gets the right Capsule").

[the remark this answers is not in the document]
Comment (jeremylenz) [reply guides/common/modules/proc_configuring-pull-transport.adoc:21]: Which setting do you mean?

`Restrict registered smart proxies` under Authentication?
`Fallback to Any Proxy` or `Enable Global Proxy` under Remote Execution?

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+After the pull-based transport is set up on each {SmartProxy}, it must also be configured on each host.
+See Configuring_and_Setting_up_Remote_Jobs_admin[Configuring and Setting Up Remote Jobs] in the _Managing Hosts_ guide.
```
[the remark this answers is not in the document]
Comment (jeremylenz) [reply guides/common/modules/proc_configuring-pull-transport.adoc]: Updated link; hopefully I did it right..

File: guides/common/modules/proc_configuring-pull-transport.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (pmoravec) [reply guides/common/modules/proc_configuring-pull-transport.adoc:21]: It is in `Settings -> Content -> Prefer registered through Capsule for remote execution`:

    Prefer using a proxy to which a host is registered when using remote execution

(relatively new(?) setting I dont see in Sat6.10, e.g.)

[the remark this answers is not in the document]
Comment (jeremylenz) [reply guides/common/modules/proc_configuring-pull-transport.adoc:21]: gotcha, thanks!

**Comment by jeremylenz** [comment]
@pmoravec @adamruzicka Updated with the setting recommendation; let me know if it's accurate..

File: guides/common/modules/con_about-running-jobs-on-hosts.adoc [hunk 8]
```diff
@@ -8,9 +8,12 @@ For custom Ansible roles that you create, or roles that you download, you must i
 Before you can use Ansible roles, you must import the roles into {Project} from the {SmartProxy} where they are installed.
 
 Communication occurs through {SmartProxyServer}, which means that {ProjectServer} does not require direct access to the target host, and can scale to manage many hosts.
-Remote execution uses the SSH service that must be enabled and running on the target host.
+
+On {SmartProxies} in `ssh` mode, remote execution uses the SSH service that must be enabled and running on the target host.
 Ensure that the remote execution {SmartProxy} has access to port 22 on the target hosts.
 
+On {SmartProxies} in `pull-mqtt` mode, remote execution uses MQTT to publish jobs it receives from {ProjectServer}, and the host subscribes to these jobs using the `yggdrasil` pull client.
+
```
Comment (pmoravec) [inline guides/common/modules/con_about-running-jobs-on-hosts.adoc:16]: Nitpick inconsistency: for `ssh` mode, we describe the requirements here ("Ensure that the remote .." 3 lines above), for `pull-mqtt` we don't describe it here as it is longer. I think it is fine as is for the sake of clarity, maybe sentence like "Requirements for the `pull-mqtt` mode are described below / in Section [link]" can be added - leaving on @jeremylenz's decision.

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 9]
```diff
@@ -0,0 +1,31 @@
+[id="configuring-pull-provider_{context}"]
+= Configuring Remote Execution for Pull Client
+
+By default, Remote Execution uses SSH as the transport mechanism for the Script provider.
+However, Remote Execution also offers a pull-based transport option, which communicates with hosts via MQTT.
+This is a good option if you are migrating from `katello-agent` which also uses pull-based communication.
+
+If you wish to use Remote Execution with the pull-based transport, run the following `{foreman-installer}` command on each {SmartProxy}:
```
Comment (pmoravec) [inline guides/common/modules/proc_configuring-pull-transport.adoc:8]: "each" -> "each relevant"? As "Some {SmartProxies} may be configured to use `pull-mqtt` mode while others use `ssh`."

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+The mode is configured per {SmartProxy}.
+Some {SmartProxies} may be configured to use `pull-mqtt` mode while others use `ssh`.
+If this is the case, it is possible that one remote job on a given host will use the pull client, and the next job on the same host will use SSH.
+If you wish to avoid this scenario, configure all {SmartProxies} to use the same mode.
+
+In `pull-mqtt` mode, hosts subscribe to the {SmartProxy} they are registered through for job notifications.
+Therefore, it is recommended to ensure that {ProjectServer} sends remote execution jobs to that same {SmartProxy}.
+To do this, in the {ProjectWebUI}, navigate to *Settings* > *Content*, and set the value of "Prefer registered through {SmartProxy} for remote execution" to *Yes*.
+
+After the pull-based transport is set up on each {SmartProxy}, it must also be configured on each host.
```
Comment (pmoravec) [inline guides/common/modules/proc_configuring-pull-transport.adoc:27]: "set up on each (proxy)" -> "set up on a (proxy)"  (as we can have ssh on another proxies, or alter one proxy from ssh to mqtt).

"on each host" -> "on each relevant host" ? To distinguish there might be hosts that will still use ssh mode?

File: guides/common/modules/proc_setting-up-pull-client.adoc [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+* The host is able to communicate with its {SmartProxy} over MQTT using port 1883.
+* The host is able to communicate with its {SmartProxy} over HTTPS.
+
+.Procedure
+. On the host, install the `katello-pull-transport-migrate` RPM package:
++
+----
+# yum install katello-pull-transport-migrate
+----
+The host's `subscription-manager` configuration and consumer certificates are used to configure the `yggdrasil` client on the host, and the worker is started.
```
Comment (pmoravec) [inline guides/common/modules/proc_setting-up-pull-client.adoc:22]: "the worker" -> "the yggdrasild worker" or "pull mode client worker"? (we speak about 'worker' for the first time)

**Comment by pmoravec** [comment]
Just minor phrasing (nitpick) issues commented, otherwise ACK to the whole PR.

File: guides/common/modules/proc_configuring-pull-transport.adoc (hunk 10 above)
Comment (jeremylenz) [reply guides/common/modules/proc_configuring-pull-transport.adoc:27]: > "set up on each (proxy)" -> "set up on a (proxy)" (as we can have ssh on another proxies, or alter one proxy from ssh to mqtt).

:+1: 

> "on each host" -> "on each relevant host" ? To distinguish there might be hosts that will still use ssh mode?

If a given (proxy) uses pull-mqtt, doesn't that mean that all of its hosts must be configured with the pull client? So I'd propose leaving "each host" as-is.

**Comment by jeremylenz** [comment]
@pmoravec thanks, updated!

**Comment by jeremylenz** [comment]
Who decides to merge this? @maximiliankolb ?

Any other changes needed?

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+This is a good option if you are migrating from `katello-agent` which also uses pull-based communication.
+
+If you wish to use Remote Execution with the pull-based transport, run the following `{foreman-installer}` command on each relevant {SmartProxy}:
+
+[options="nowrap" subs="quotes,attributes"]
+----
+
+ # {installer-scenario-smartproxy} \
+  --foreman-proxy-plugin-remote-execution-script-mode pull-mqtt
+
```
Comment (maximiliankolb) [inline guides/common/modules/proc_configuring-pull-transport.adoc:15]:
```suggestion
```

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+However, Remote Execution also offers a pull-based transport option, which communicates with hosts via MQTT.
+This is a good option if you are migrating from `katello-agent` which also uses pull-based communication.
+
+If you wish to use Remote Execution with the pull-based transport, run the following `{foreman-installer}` command on each relevant {SmartProxy}:
+
+[options="nowrap" subs="quotes,attributes"]
+----
+
+ # {installer-scenario-smartproxy} \
+  --foreman-proxy-plugin-remote-execution-script-mode pull-mqtt
```
Comment (maximiliankolb) [inline guides/common/modules/proc_configuring-pull-transport.adoc:14]:
```suggestion
--foreman-proxy-plugin-remote-execution-script-mode pull-mqtt
```

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 14]
```diff
@@ -0,0 +1,31 @@
+[id="configuring-pull-provider_{context}"]
+= Configuring Remote Execution for Pull Client
+
+By default, Remote Execution uses SSH as the transport mechanism for the Script provider.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_configuring-pull-transport.adoc:4]: what do you mean by "script provider"?

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+However, Remote Execution also offers a pull-based transport option, which communicates with hosts via MQTT.
+This is a good option if you are migrating from `katello-agent` which also uses pull-based communication.
+
+If you wish to use Remote Execution with the pull-based transport, run the following `{foreman-installer}` command on each relevant {SmartProxy}:
+
```
Comment (maximiliankolb) [inline guides/common/modules/proc_configuring-pull-transport.adoc:9]:
```suggestion
.Procedure
* Enable the pull-based transport on your {SmartProxyServer}:
```

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+[options="nowrap" subs="quotes,attributes"]
+----
+
```
Comment (maximiliankolb) [inline guides/common/modules/proc_configuring-pull-transport.adoc:12]:
```suggestion
```

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+ # {installer-scenario-smartproxy} \
+  --foreman-proxy-plugin-remote-execution-script-mode pull-mqtt
+
+----
+
+The mode is configured per {SmartProxy}.
+Some {SmartProxies} may be configured to use `pull-mqtt` mode while others use `ssh`.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_configuring-pull-transport.adoc:19]: This sounds a bit weird in a procedure to configure the pull transport.

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+If this is the case, it is possible that one remote job on a given host will use the pull client, and the next job on the same host will use SSH.
+If you wish to avoid this scenario, configure all {SmartProxies} to use the same mode.
+
+In `pull-mqtt` mode, hosts subscribe to the {SmartProxy} they are registered through for job notifications.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_configuring-pull-transport.adoc:23]: This paragraph sounds like a different procedure.

File: guides/common/modules/proc_distributing-keys-for-remote-execution.adoc [hunk 19]
```diff
@@ -1,7 +1,9 @@
 [id="Distributing_SSH_Keys_for_Remote_Execution_{context}"]
 = Distributing SSH Keys for Remote Execution
 
-To use SSH keys for authenticating remote execution connections, you must distribute the public SSH key from {SmartProxy} to its attached hosts that you want to manage.
+For {SmartProxies} in `ssh` mode, remote execution connections are authenticated using SSH.
+In most cases, the public SSH key from {SmartProxy} must be distributed to its attached hosts that you want to manage.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_distributing-keys-for-remote-execution.adoc:5]: What do you mean by "most cases"?

File: guides/common/modules/proc_migrating-from-katello-agent-to-remote-execution.adoc [hunk 20]
```diff
@@ -15,6 +15,10 @@ endif::[]
 * You have previously installed the `katello-agent` package on content hosts.
 
 .Procedure
+. _If you have Remote Execution configured to use `pull-mqtt` mode_: Deploy the remote execution pull client to the hosts.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_migrating-from-katello-agent-to-remote-execution.adoc:18]:
```suggestion
. If you have Remote Execution configured to use `pull-mqtt` mode: Deploy the remote execution pull client to the hosts.
```

Maybe reword to "Ensure package X is installed on your managed host"?

File: guides/common/modules/proc_migrating-from-katello-agent-to-remote-execution.adoc [hunk 21]
```diff
@@ -15,6 +15,10 @@ endif::[]
...
+Follow the procedure in xref:setting-up-pull-client_{context}[].
```
Comment (maximiliankolb) [inline guides/common/modules/proc_migrating-from-katello-agent-to-remote-execution.adoc:19]:
```suggestion
For more information, see xref:setting-up-pull-client_{context}[].
```

File: guides/common/modules/proc_migrating-from-katello-agent-to-remote-execution.adoc [hunk 22]
```diff
@@ -15,6 +15,10 @@ endif::[]
...
+. _If you have Remote Execution configured to use `ssh` mode_: Distribute the remote execution SSH keys to the hosts.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_migrating-from-katello-agent-to-remote-execution.adoc:20]: same as above.

File: guides/common/modules/proc_setting-up-pull-client.adoc [hunk 23]
```diff
@@ -0,0 +1,34 @@
+[id="setting-up-pull-client_{context}"]
```
Comment (maximiliankolb) [inline guides/common/modules/proc_setting-up-pull-client.adoc:1]: title does not match anchor and file name. Please adjust this.

File: guides/common/modules/proc_setting-up-pull-client.adoc [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+= Setting Up the Remote Execution Pull Client
+
+For {SmartProxies} in `pull-mqtt` mode, hosts can subscribe to remote jobs using the remote execution pull client.
+No SSH connection is required.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_setting-up-pull-client.adoc:5]:
```suggestion
Managed hosts do not require an SSH connection to their {SmartProxyServer}.
```

File: guides/common/modules/proc_setting-up-pull-client.adoc [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+
+Use this procedure to configure a host to use the pull client:
```
Comment (maximiliankolb) [inline guides/common/modules/proc_setting-up-pull-client.adoc:7]: This line does not really match the title. Maybe reword the title and drop this line.

File: guides/common/modules/proc_setting-up-pull-client.adoc [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+
+.Prerequisites
+* The host is registered to {ProjectName}.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_setting-up-pull-client.adoc:10]: Maybe reword to "You have registered the host to {Project}."?

File: guides/common/modules/proc_setting-up-pull-client.adoc [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+* The host's {SmartProxy} is in `pull-mqtt` mode.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_setting-up-pull-client.adoc:11]: Please rephrase.

File: guides/common/modules/proc_setting-up-pull-client.adoc [hunk 28]
```diff
@@ -0,0 +1,34 @@
...
+* The {Team} {project-client-name} repository is enabled and synchronized on {ProjectServer}, and enabled on the host.
+* The host is able to communicate with its {SmartProxy} over MQTT using port 1883.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_setting-up-pull-client.adoc:13]:
```suggestion
* The host is able to communicate with its {SmartProxy} over MQTT using port `1883`.
```

File: guides/common/modules/proc_setting-up-pull-client.adoc [hunk 29]
```diff
@@ -0,0 +1,34 @@
...
+* The host is able to communicate with its {SmartProxy} over HTTPS.
+
+.Procedure
+. On the host, install the `katello-pull-transport-migrate` RPM package:
++
+----
+# yum install katello-pull-transport-migrate
```
Comment (maximiliankolb) [inline guides/common/modules/proc_setting-up-pull-client.adoc:20]: This seems to depend on the OS/EL version. what about el8? debian? Is this EL only?

File: guides/common/modules/proc_setting-up-pull-client.adoc [hunk 30]
```diff
@@ -0,0 +1,34 @@
...
+
+.Prerequisites
+* The host is registered to {ProjectName}.
+* The host's {SmartProxy} is in `pull-mqtt` mode.
+* The {Team} {project-client-name} repository is enabled and synchronized on {ProjectServer}, and enabled on the host.
+* The host is able to communicate with its {SmartProxy} over MQTT using port 1883.
+* The host is able to communicate with its {SmartProxy} over HTTPS.
+
+.Procedure
+. On the host, install the `katello-pull-transport-migrate` RPM package:
```
Comment (maximiliankolb) [inline guides/common/modules/proc_setting-up-pull-client.adoc:17]:
```suggestion
. Install the `katello-pull-transport-migrate` package on your managed host:
```

**Review by maximiliankolb — CHANGES_REQUESTED** [review]
Thanks for your PR. I made some comments, but I could not test this yet.

File: guides/common/modules/proc_configuring-pull-transport.adoc (hunk 14 above)
Comment (jeremylenz) [reply guides/common/modules/proc_configuring-pull-transport.adoc:4]: Remote execution has several providers - Script (formerly SSH) and Ansible being the most common. The provider decides which job template is used.

File: guides/common/modules/proc_setting-up-pull-client.adoc (hunk 29 above)
Comment (jeremylenz) [reply guides/common/modules/proc_setting-up-pull-client.adoc:20]: `subscription-manager` is required, so I'd say yes it's EL-only.

File: guides/common/modules/proc_configuring-pull-transport.adoc (hunk 14 above)
Comment (maximiliankolb) [reply guides/common/modules/proc_configuring-pull-transport.adoc:4]: Just checked on Foreman 3.3 that "Script" is new. So capitalizing it is IMHO very much correct. 👍

File: guides/common/modules/proc_setting-up-pull-client.adoc (hunk 29 above)
Comment (maximiliankolb) [reply guides/common/modules/proc_setting-up-pull-client.adoc:20]: FYI: You can also run subscription-manager on sles and Debian/Ubuntu: https://apt.atix.de/

So maybe make the "yum" conditional?

Comment (jeremylenz) [reply guides/common/modules/proc_setting-up-pull-client.adoc:20]: @adamruzicka @ehelms Is the pull client supported on Debian?

File: guides/common/modules/proc_distributing-keys-for-remote-execution.adoc (hunk 19 above)
Comment (jeremylenz) [reply guides/common/modules/proc_distributing-keys-for-remote-execution.adoc:5]: Below there's a note "If the hosts are running on Amazon Web Services, enable password authentication." If you're using password authentication,  I think that would be the exception to "most cases."  (check me @adamruzicka )

File: guides/common/modules/proc_setting-up-pull-client.adoc (hunk 29 above)
Comment (adamruzicka) [reply guides/common/modules/proc_setting-up-pull-client.adoc:20]: I'd be inclined to say no. The binaries we build are static and should run essentially anywhere as long as the platform and architecture match, but afaik we don't build packages for debian.

File: guides/common/modules/proc_distributing-keys-for-remote-execution.adoc (hunk 19 above)
Comment (adamruzicka) [reply guides/common/modules/proc_distributing-keys-for-remote-execution.adoc:5]: This feels a bit like playing with words. To be pedantic, you don't *have to* distribute the keys, but if you want to use REX, you have to make it work somehow. Using keys just happens to be the preferred way.

File: guides/common/modules/proc_setting-up-pull-client.adoc (hunk 29 above)
Comment (maximiliankolb) [reply guides/common/modules/proc_setting-up-pull-client.adoc:20]: Thanks for the info. I guess the suggestion remains to split the "yum install" command into EL7 and EL8/9 (dnf).

File: guides/common/modules/proc_configuring-pull-transport.adoc (hunk 18 above)
Comment (jeremylenz) [reply guides/common/modules/proc_configuring-pull-transport.adoc:21]: See [above](https://github.com/theforeman/foreman-documentation/pull/1548/files#r961579453) for some context. I agree it does kind of feel like a different procedure, but it should be done at the same time you configure your {SmartProxies} for pull mode. Given that, what's the best place for this text?

File: guides/common/modules/proc_setting-up-pull-client.adoc (hunk 23 above)
Comment (jeremylenz) [reply guides/common/modules/proc_setting-up-pull-client.adoc:1]: sorry, I'm not very good at asciidoc yet. Could you please clarify what I need to change to fix this?

Comment (maximiliankolb) [reply guides/common/modules/proc_setting-up-pull-client.adoc:1]: Absolutely, thanks for asking! In general, we ensure that the file name, the anchor, and the title "match" (minus attributes and the type prefix such as `con_` or `proc_`).

For your example:

Anchor: `[id="Setting_Up_the_Remote_Execution_Pull_Client_{context}"]` (without prefix, with `{context}` suffix)
Title: `Setting Up the Remote Execution Pull Client` (original)
File name: `proc_setting-up-the-remote-execution-pull-client.adoc` (all lowercase; only single underscore)

If you want to adjust the title, then I'd kindly ask you to also adjust the anchor and file name.

Comment (jeremylenz) [reply guides/common/modules/proc_setting-up-pull-client.adoc]: Changed file name and anchor so they match the new title.

**Comment by jeremylenz** [comment]
@maximiliankolb Updated, thanks :+1:

File: guides/common/modules/proc_configuring-pull-transport.adoc (hunk 18 above)
Comment (maximiliankolb) [reply guides/common/modules/proc_configuring-pull-transport.adoc:22]: You can potentially split the procedure into two. But IMHO this can be done at a later stage.

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 31]
```diff
@@ -0,0 +1,30 @@
+[id="configuring-pull-provider_{context}"]
+= Configuring Remote Execution for Pull Client
```
Comment (maximiliankolb) [inline guides/common/modules/proc_configuring-pull-transport.adoc:2]: Title does not match anchor and file name.

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+
+By default, Remote Execution uses SSH as the transport mechanism for the Script provider.
+However, Remote Execution also offers a pull-based transport option, which communicates with hosts via MQTT.
+This is a good option if you are migrating from `katello-agent` which also uses pull-based communication.
+
+.Procedure
+	* Enable the pull-based transport on each relevant {SmartProxyServer}:
```
Comment (maximiliankolb) [inline guides/common/modules/proc_configuring-pull-transport.adoc:9]:
```suggestion
* Enable the pull-based transport on each relevant {SmartProxyServer}:
```

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 33]
```diff
@@ -0,0 +1,30 @@
...
+
```
Comment (maximiliankolb) [inline guides/common/modules/proc_configuring-pull-transport.adoc:10]:
```suggestion
+
```

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 34]
```diff
@@ -0,0 +1,30 @@
...
+----
+
+The mode is configured per {SmartProxy}.
+Some {SmartProxies} may be configured to use `pull-mqtt` mode while others use `ssh`.
+If this is the case, it is possible that one remote job on a given host will use the pull client, and the next job on the same host will use SSH.
+If you wish to avoid this scenario, configure all {SmartProxies} to use the same mode.
+
+In `pull-mqtt` mode, hosts subscribe to the {SmartProxy} they are registered through for job notifications.
+Therefore, it is recommended to ensure that {ProjectServer} sends remote execution jobs to that same {SmartProxy}.
+To do this, in the {ProjectWebUI}, navigate to *Administer* > *Settings* > *Content*, and set the value of "Prefer registered through {SmartProxy} for remote execution" to *Yes*.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_configuring-pull-transport.adoc:24]:
```suggestion
To do this, in the {ProjectWebUI}, navigate to *Administer* > *Settings*.
On the *Content* tab, set the value of "Prefer registered through {SmartProxy} for remote execution" to *Yes*.
```

File: guides/common/modules/proc_configuring-pull-transport.adoc [hunk 35]
```diff
@@ -0,0 +1,30 @@
...
+
+After the pull-based transport is set up on a {SmartProxy}, it must also be configured on each host.
+See {ManagingHostsDocURL}Configuring_and_Setting_Up_Remote_Jobs_managing-hosts[Configuring and Setting Up Remote Jobs] in the _Managing Hosts Guide_
```
Comment (maximiliankolb) [inline guides/common/modules/proc_configuring-pull-transport.adoc:27]:
```suggestion
For more information, see {ManagingHostsDocURL}Configuring_and_Setting_Up_Remote_Jobs_managing-hosts[Configuring and Setting Up Remote Jobs] in _{ManagingHostsDocTitle}_.
```

**Review by maximiliankolb — COMMENTED** [review]
A few minor suggestions, rest LGTM!

**Review by maximiliankolb — APPROVED** [review]
LGTM!

**Comment by maximiliankolb** [comment]
handing over to @Lennonka Please squash when merging.

**Comment by Lennonka** [comment]
Hi, sorry, but can you please help me verify that I'm not duplicating this in #1630 ?

**Comment by jeremylenz** [comment]
yes, I had mentioned this in the previous PR: https://github.com/theforeman/foreman-documentation/pull/1584#issuecomment-1239468663

I will comment on https://github.com/theforeman/foreman-documentation/pull/1630.

**Comment by Lennonka** [comment]
FYI, I will cherry-pick this to both 3.3 and 3.4, because for Satellite 6.12, we need it in the 3.3 branch.

**Comment by Lennonka** [comment]
Cherry-picked:
- b1940f9e8..96d9d7417  3.4 -> 3.4
- 41a14e2ce..ae4dab747  3.3 -> 3.3

Note: I forgot to squash the commits during the merge in master, so I squashed them for 3.3 and 3.4.
